File: scala/src/main/scala/folio/codinginterview/presentation/MarketPriceController.scala
package folio.codinginterview.presentation

import folio.codinginterview.application.usecase.market_price.UpdateMarketPriceItemInput
import folio.codinginterview.application.usecase.market_price.UpdateMarketPriceUsecase
import folio.codinginterview.application.usecase.market_price.UpdateMarketPriceUsecaseInput
import folio.codinginterview.domain.StockSymbol
import folio.codinginterview.presentation.PresentationException.BadRequestException
import scala.concurrent.Future

object MarketPriceController {
  final case class MarketPriceItemDto(symbol: String, market_price: String)
  final case class UpdateMarketPriceRequest(market_prices: Seq[MarketPriceItemDto])
}

final class MarketPriceController(
    updateMarketPriceUsecase: UpdateMarketPriceUsecase
) {
  import MarketPriceController.*

  def updateMarketPrice(req: UpdateMarketPriceRequest): Future[Unit] = {
    val parsed: Either[String, Seq[UpdateMarketPriceItemInput]] =
      req.market_prices.foldLeft[Either[String, Seq[UpdateMarketPriceItemInput]]](Right(Seq.empty)) { case (acc, dto) =>
        for {
          xs <- acc
          sym <- StockSymbol
            .fromString(dto.symbol)
            .toRight(s"unknown symbol: ${dto.symbol}")
          price <-
            try Right(BigDecimal(dto.market_price))
            catch { case _: Throwable => Left(s"invalid market_price: ${dto.market_price}") }
        } yield xs :+ UpdateMarketPriceItemInput(sym, price)
      }
    parsed match {
      case Left(msg)    => Future.failed(BadRequestException(msg))
      case Right(items) => updateMarketPriceUsecase.run(UpdateMarketPriceUsecaseInput(items))
    }
  }
}

File: typescript/src/presentation/marketPriceController.ts
import Decimal from "decimal.js";
import {
  UpdateMarketPriceItemInput,
  UpdateMarketPriceUsecase,
} from "../application/usecase/market_price/updateMarketPriceUsecase.js";
import { StockSymbol } from "../domain/stockSymbol.js";
import { BadRequestException } from "./presentationException.js";

export interface MarketPriceItemDto {
  symbol: string;
  market_price: string;
}

export interface UpdateMarketPriceRequest {
  market_prices: MarketPriceItemDto[];
}

export class MarketPriceController {
  constructor(private readonly updateMarketPriceUsecase: UpdateMarketPriceUsecase) {}

  async updateMarketPrice(req: UpdateMarketPriceRequest): Promise<void> {
    const items: UpdateMarketPriceItemInput[] = [];
    for (const dto of req.market_prices) {
      const sym = StockSymbol.fromString(dto.symbol);
      if (sym === undefined) {
        throw new BadRequestException(`unknown symbol: ${dto.symbol}`);
      }
      let price: Decimal;
      try {
        price = new Decimal(dto.market_price);
      } catch {
        throw new BadRequestException(`invalid market_price: ${dto.market_price}`);
      }
      items.push({ symbol: sym, marketPrice: price });
    }
    await this.updateMarketPriceUsecase.run({ items });
  }
}
